### ISSUE TITLE
segmentation fault when using Tensor.numpy() and Tensor.from_numpy() in backward()

### ISSUE THREAD

**Issue by jeasinema** [pytorch/pytorch#3219]
The code below will cause a segmentation fault on some test platforms:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.autograd import Variable
import numpy as np


class Foo(torch.autograd.Function):
    def forward(ctx, x):
        return torch.from_numpy(x.numpy())
    def backward(ctx, grad_output):
        a = grad_output.numpy()
        # a = a + 0.
        return torch.from_numpy(a)

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(16, 16)
        self.fc2 = nn.Linear(16, 16)
        self.foo = Foo()

    def forward(self, x):
        x = self.fc1(x)
        x = self.foo(x)
        # x = self.fc2(x)
        x = x.view(1, 16)
        return x;

def train(epoch):
    model = Net()
    model.train()
    optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.99)
    data = Variable(torch.from_numpy(np.ones((1, 16), dtype=np.float32))) #TODO
    target = Variable(torch.from_numpy(np.zeros((1), dtype=np.int64))) #TODO

    test_data = torch.from_numpy(np.ones((1), dtype=np.float32)) #TODO
    def func(a):
        b = a.numpy()
        # b = b + 0.
        b[0] = 0
        return torch.from_numpy(b)
    print(test_data)
    d = func(test_data)
    print(d)
    print(test_data)
    d[0] = 2
    print(d)
    print(test_data)

    optimizer.zero_grad()
    output = model(data)
    loss = F.nll_loss(output, target)
    loss.backward()
    optimizer.step()

def test():
    model = torch.load(args.model_path)
    model.eval()

def main():
    train(0)

if __name__ == "__main__":
    main()
```
Some useful information:
1. Test platform 1:
- `Ubuntu 16.04.3`
- `gcc 5.4.0`
- `Python 3.5.2`
- `PyTorch 0.2.0_3`
2. Test platform 2:
- `macOS 10.12.6`
- `Apple LLVM version 9.0.0 (clang-900.0.38)`
- `Python 3.6.3`
- `PyTorch 0.2.0_2`
3. If the customized layer is placed the first layer in the model, this error will not happen since there is no gradient for it to back prop to the layer before. 
4. Compared to the operation in `def func(a)`, the code here will be executed in "pure python", however, the back prop code are connected by PyTorch's C backend. But I'm not familiar with this part.
5. If I uncomment the `a = a + 0`, this error will disappear.

Any idea on this problem?

**Comment by zou3519** [comment]
This might be related to this: https://discuss.pytorch.org/t/torch-from-numpy-in-custom-backward-doesnt-work/9791/18

If it is the same issue, then building from source or upgrading to 0.3 when it comes out should fix the problem.
